### PR TITLE
Added more functionalities to OptimizedRNN operator

### PR DESCRIPTION
**Pull request by yuanbyu**
Added two new cell types: 
 - lstm-peepholes: Do peephole in LSTM. Peephole weight is an additional input.
 - lstm-input-forget: Couple the input and forget gates.

Added the following attributes:
 - activations: A list of activation functions for the gates, so we could customize the activations.
 - use_bias: Determine whether to use the bias tensor. Default to true.
 - reverse: Reverse the inputs.  Default is false.
 - clip: Cell clip threshold. Default to no clip.

The weight packing in input 0 seems to be under-specified. For example, in the lstm case, it is unclear how Wi, Wf, Wo, and Wc are packed in the weight tensor. I can try to improve it if I know the layout.

### CODE REVIEW

File: onnx/defs/rnn/defs.cc [hunk 1]
```diff
@@ -108,18 +136,19 @@ The order of matrixes `{K, L, D, R, N, C}` is defined as:
...
            "tensor with the shape of `[seq_length, batch_size, input_size]`.")
-    // TODO: do we want to allow different lengths of sequences in a minibatch?
-    // CuDNN supports it, but not all backend implementations do. One way to
-    // encode would be int-valued tensor denoting lengths of each sequence in
-    // the batch.
     .Input(2, "initial_h",
            "Optional initial value of the hidden. If not specified - assumed "
-           "to be 0. Dimensions `[num_layers * directions, batch_size, "
-           "hidden_size]`")
+           "to be 0. Dimensions `[num_layers * directions, hidden_size]`")
```
Comment (dzhulgakov) [inline onnx/defs/rnn/defs.cc]: actually, why did you remove `batch_size`? Cudnn allows to initialize different state per example and it in general makes sense (if you want to continue previous sequence)

File: onnx/defs/rnn/defs.cc [hunk 2]
```diff
@@ -108,18 +136,19 @@ The order of matrixes `{K, L, D, R, N, C}` is defined as:
...
     .Input(3, "initial_c",
            "For LSTM only: optional initial value of the cell. If not "
            "specified - assumed to be 0. Dimensions `[num_layers * directions, "
-           "batch_size, hidden_size]`")
+           "hidden_size]`")
+    .Input(4, "seq_lens",
+           "Optional tensor specifying lengths of the sequences in a batch."
+           "Has shape `[batch_size]`.")
+    .Input(5, "peephole_w",
```
Comment (dzhulgakov) [inline onnx/defs/rnn/defs.cc]: shouldn't we merge it into the global block of weights above?

File: onnx/defs/rnn/defs.cc [hunk 3]
```diff
@@ -71,6 +90,15 @@ allows to make spec more uniform.
     .Attr("num_layers", "Numbers of RNN layers in the stack, default 1",
           AttrType::INT)
     .Attr("hidden_size", "Number of neurons in the hidden layer", AttrType::INT)
+    .Attr("activations", "A list of activation functions for the gates. Typical activation "
```
Comment (dzhulgakov) [inline onnx/defs/rnn/defs.cc]: you need to describe the order of activation functions (probably the same as the gates order in weights blob below). And say that default version just assumes stuff from the formula above.

File: onnx/defs/rnn/defs.cc [hunk 4]
```diff
@@ -71,6 +90,15 @@ allows to make spec more uniform.
...
+          "functions are sigmoid and tanh",
+          AttrType::STRINGS)
+    .Attr("use_bias", "Use the bias tensors for the gates if 1. Default 1.",
```
Comment (dzhulgakov) [inline onnx/defs/rnn/defs.cc]: edit the comment for weights to describe that bias is optional in this case?

File: onnx/defs/rnn/defs.cc [hunk 5]
```diff
@@ -71,6 +90,15 @@ allows to make spec more uniform.
...
+          AttrType::INT)
+    .Attr("reverse", "Process the sequences in reverse order, default 0",
+          AttrType::INT)
+    .Attr("clip", "Cell clip threshold. Default to no clip if not specified.",
```
Comment (dzhulgakov) [inline onnx/defs/rnn/defs.cc]: add more comments in the spec above regarding where exactly 'clip' gets plugged in?

File: onnx/defs/rnn/defs.cc [hunk 6]
```diff
@@ -82,7 +110,7 @@ Each parameter matrix is linearly appended after the previous parameter matrix
 without padding.
 
 The order of matrixes `{K, L, D, R, N, C}` is defined as:
```
Comment (dzhulgakov) [inline onnx/defs/rnn/defs.cc]: you'd need to update the comment here too - for example change 'relu' / 'tanh' to 'simple' and update layout order

File: onnx/defs/rnn/defs.cc [hunk 7]
```diff
@@ -108,18 +136,19 @@ The order of matrixes `{K, L, D, R, N, C}` is defined as:
...
            "Optional initial value of the hidden. If not specified - assumed "
-           "to be 0. Dimensions `[num_layers * directions, batch_size, "
-           "hidden_size]`")
+           "to be 0. Dimensions `[num_layers * directions, hidden_size]`")
     .Input(3, "initial_c",
            "For LSTM only: optional initial value of the cell. If not "
            "specified - assumed to be 0. Dimensions `[num_layers * directions, "
-           "batch_size, hidden_size]`")
+           "hidden_size]`")
+    .Input(4, "seq_lens",
```
Comment (dzhulgakov) [inline onnx/defs/rnn/defs.cc]: with several optional args the ordering is unfortunate. Maybe we should allow for 'None' (or empty string) in inputs for this purpose

File: onnx/defs/rnn/defs.cc (hunk 2 above)
Comment (yuanbyu) [reply onnx/defs/rnn/defs.cc]: This is related to the layout of the weights: We may have to pad if the layout is not right. Also, it is optional. So I made it a separate input.  It is fine with me to pack it into weights.

File: onnx/defs/rnn/defs.cc (hunk 7 above)
Comment (yuanbyu) [reply onnx/defs/rnn/defs.cc]: Reverted.

File: onnx/defs/rnn/defs.cc (hunk 1 above)
Comment (yuanbyu) [reply onnx/defs/rnn/defs.cc]: I don't know it. Reverted.

File: onnx/defs/rnn/defs.cc (hunk 6 above)
Comment (yuanbyu) [reply onnx/defs/rnn/defs.cc]: Thanks. Done.
Regarding the layout order in the weights input, I don't have a complete picture. Take a simple example, for LSTM, it is not completely clear to me how we pack Wi, Wf, Wo, and Wc in the weight tensor. Is it [4*hidden_size, input_size] or [hidden_size, 4*input_size]?  Perhaps I missed something obvious.

File: onnx/defs/rnn/defs.cc (hunk 5 above)
Comment (yuanbyu) [reply onnx/defs/rnn/defs.cc]: Done.

File: onnx/defs/rnn/defs.cc (hunk 4 above)
Comment (yuanbyu) [reply onnx/defs/rnn/defs.cc]: Biases are still included in input(0) but not used if use_bias is false.

File: onnx/defs/rnn/defs.cc (hunk 3 above)
Comment (yuanbyu) [reply onnx/defs/rnn/defs.cc]: Done.

File: docs/Operators.md [hunk 8]
```diff
@@ -885,47 +887,72 @@ Equation definitions:
...
+  - it = sigmoid(Wi*Xt + Ri*Ht-1 + Pi(.)Ct-1 + Wbi + Rbi)
+  - ft = sigmoid(Wf*Xt + Rf*Ht-1 + Pf(.)Ct-1 + Wbf + Rbf)
+  - ot = sigmoid(Wo*Xt + Ro*Ht-1 + Po(.)Ct-1 + Wbo + Rbo)
+  - ct = tanh(Wc*Xt + Rc*Ht-1 + Wbc + Rbc)
+  - Ct = ft (.) Ct-1 + it (.) ct
+  - H = ot (.) tanh(C)
+
+  `lstm_input_forget with default activations`
+  - it = sigmoid(Wi*Xt + Ri*Ht-1 + Wbi + Rbi)
+  - ft = sigmoid(Wf*Xt + Rf*Ht-1 + Wbf + Rbf)
```
Comment (raduk) [inline docs/Operators.md]: This is not needed anymore. We may also comment that forget weights are ignored for this type of cell

File: docs/Operators.md [hunk 9]
```diff
@@ -885,47 +887,72 @@ Equation definitions:
...
+`Rb[izrfcohp]` - R parameter bias vectors for the corresponding gates
 `ReLU(X)` - max(X, 0)
 `tanh` - hyperbolic tangent of X
 `sigmoid(X)` - 1 / (1 + e^-X)
 `[C|H]` - Cell/Hidden state
 
 - Equations:
-  `relu`
+  `simple(relu)`
   - Ht = ReLU(Wi*Xt + Ri*Ht-1 + Wbi + Rbi)
```
Comment (raduk) [inline docs/Operators.md]: We may add the clip to all the formulas: ReLU(clip(Wi*Xt + Ri*Ht-1 + Wbi + Rbi))

Comment (yuanbyu) [reply docs/Operators.md]: Clip is optional and disabled as default. There are comments on the clip attribute.

File: docs/Operators.md (hunk 8 above)
Comment (yuanbyu) [reply docs/Operators.md]: Good catch. Removed.

**Comment by yuanbyu** [comment]
@dzhulgakov As I understand it, this one is blocked due to the optional inputs issue #151?

File: onnx/defs/rnn/defs.cc (hunk 6 above)
Comment (dzhulgakov) [reply onnx/defs/rnn/defs.cc]: It's actually more complicated as we're trying to pack several layers of e.g. LSTM together. The first layer would have different input size from later ones, the input gates might not have inputs at all, etc. Right now I just concatenated all the weight matrices together. It's not pretty but solves the problem of too many inputs.

Thus for peep_hole connection I'd just modify the message to say that there are more weight matrices in the internal order (R being "input-hidden", "hidden-hidden" and "peephole").

**Comment by dzhulgakov** [comment]
Yep, let me hack #151.

In the meantime we also need to figure out what to do with peephole connections weights. If you have better suggestions on how to layout the weight matrices - please share.

File: onnx/defs/rnn/defs.cc (hunk 6 above)
Comment (yuanbyu) [reply onnx/defs/rnn/defs.cc]: @dzhulgakov As we discussed, I will try to update this PR to have 4 RNN ops: SimpleRNN, GRU, LSTM, and BirectionalLSTM. I think that this would make things simpler.

Comment (dzhulgakov) [reply onnx/defs/rnn/defs.cc]: Yes, adding separate ops would make it simpler. The only problem is that stacked nature - for example the input-hidden matrix has different sizes for the first and subsequent layers.

**Comment by yuanbyu** [comment]
@dzhulgakov @raduk @ebarsoum I have updated this PR as we discussed:

1. Removed OptimizedRNN and replaced it with 4 simpler ops: SimpleRNN, GRU, LSTM, BidirectionalLSTM. 
2. The current version doesn't support multi-layer RNNs. We will add it in a separate PR with a new version number.

File: onnx/defs/rnn/defs.cc [hunk 10]
```diff
@@ -5,124 +5,311 @@
...
- -- For other layers (`L>1`) weight matrix (`K=weight`) on input connection
-    (`R=input-hidden`), dimensions are `{hidden_size, directions * hidden_size}`
- -- For weight matrix (`K=weight`) on recurrent connection (`R=hidden-hidden`),
-    dimensions are `{hidden_size, hidden_size}`
- -- For all biases (`K=bias`), dimensions are `{hidden_size}`
+    .Attr("reverse", "Process the sequences in reverse order if 1, default 0.",
+          AttrType::INT)
+    .Attr("input_forget", "Couple the input and forget gates if 1, default 0.",
+          AttrType::INT)	  
+    .Attr("clip", "Cell clip threshold. Clipping bounds the elements of a tensor "
```
Comment (dzhulgakov) [inline onnx/defs/rnn/defs.cc]: for repetitive arguments like this you may find FillUsing useful and extract common args to a subfunction. See for example https://github.com/onnx/onnx/blob/ea6d6578d9d2e566be69077a8c5b43663b0649c7/onnx/defs/reduction/defs.cc#L36

File: onnx/defs/rnn/defs.cc [hunk 11]
```diff
@@ -5,124 +5,311 @@
...
+  - Ct = ft (.) Ct-1 + it (.) ct
+  - H = ot (.) tanh(Ct)
 )DOC")
-    .Input(1, "input",
+    .Attr("activations", "A list of 4 activation functions for input, output, "
+	  "forget, and cell gates. Typical activation functions are sigmoid "
+	  "and tanh. See the equations for default if not specified.",
+          AttrType::STRINGS)
+    .Attr("hidden_size", "Number of neurons in the hidden layer", AttrType::INT)
+    .Attr("input_forget", "Couple the input and forget gates if 1, default 0.",
```
Comment (dzhulgakov) [inline onnx/defs/rnn/defs.cc]: so if you couple them the input weight would be different? should we instead create a yet another op instead?

File: onnx/defs/rnn/defs.cc [hunk 12]
```diff
@@ -5,124 +5,311 @@
...
+`Wbi` - W parameter bias vector for input gate
+`Rbi` - R parameter bias vector for input gate
+`ReLU(X)` - max(X, 0)
+`tanh(X)` - hyperbolic tangent of X
+`H` - Hidden state
+
+Equations:
+  - Ht = Activation(Wi*Xt + Ri*Ht-1 + Wbi + Rbi)
+)DOC")
+    .Attr("activation", "The activation function for input gate. Typical "
```
[the remark this answers is not in the document]
Comment (yuanbyu) [reply onnx/defs/rnn/defs.cc]: Thanks. Done.

File: onnx/defs/rnn/defs.cc [hunk 13]
```diff
@@ -5,124 +5,311 @@
...
+    .Attr("reverse", "Process the sequences in reverse order, default 0.",
+          AttrType::INT)
+    .Attr("clip", "Cell clip threshold. Clipping bounds the elements of a tensor "
+          "in the range of [-threshold, +threshold] and is applied to the input "
+          "of activations. No clip if not specified.",
+          AttrType::FLOAT)
+    .Input(0, "input",
+           "The input sequences packed (and potentially padded) into one 3-D "
+           "tensor with the shape of `[seq_length, batch_size, input_size]`.")
+    .Input(1, "W",
```
[the remark this answers is not in the document]
Comment (yuanbyu) [reply onnx/defs/rnn/defs.cc]: I would prefer to keep it not optional for this PR, and do it in the next one.

File: onnx/defs/rnn/defs.cc [hunk 14]
```diff
@@ -5,124 +5,311 @@
...
+	   "`[hidden_size, input_size]`.")
+    .Input(2, "R",
+	   "The recurrence weight tensor. The tensor has shape "
+	   "`[hidden_size, hidden_size}`.")
+    .Input(3, "B",
+	   "The bias tensor for input gate. The tensor is a concatenation of"
+	   "`Wbi` and `Rbi`, and has shape `[2*hidden_size]`, Optional: If not "
+	   "specified - assumed to be 0.",
+	   true /*optional*/)
+    .Input(4, "initial_h",
```
[the remark this answers is not in the document]
Comment (yuanbyu) [reply onnx/defs/rnn/defs.cc]: I have a slight preference to keep it simple as is.

File: onnx/defs/rnn/defs.cc [hunk 15]
```diff
@@ -5,124 +5,311 @@
...
+          "in the range of [-threshold, +threshold] and is applied to the input "
+          "of activations. No clip if not specified.",
+          AttrType::FLOAT)
+    .Input(0, "input",
+           "The input sequences packed (and potentially padded) into one 3-D "
+           "tensor with the shape of `[seq_length, batch_size, input_size]`.")
+    .Input(1, "W",
+	   "The weight tensor for the gates. The weights W[zrh] are "
+	   "concatenated along dimension 0. This tensor has shape "
+	   "`[3 * hidden_size, input_size]`.")
```
[the remark this answers is not in the document]
Comment (yuanbyu) [reply onnx/defs/rnn/defs.cc]: It seems that both CoreML and PyTorch use this notation so I will keep it as is.

File: onnx/defs/rnn/defs.cc [hunk 16]
```diff
@@ -5,124 +5,311 @@
...
+    .Input(2, "R",
+	   "The recurrence weight tensor. The recurrence weights R[zrh] are "
+	   "concatenated along dimension 0. This tensor has shape "
+	   "`[3 * hidden_size, hidden_size}`.")
+    .Input(3, "B",
+	   "The bias tensor for the gates. The biases Wb[zrh] and Rb[zrh] are "
+	   "concatenated along dimension 0. This tensor has shape "
+	   "`[6 * hidden_size]`. Optional: If not specified - assumed to be 0",
```
[the remark this answers is not in the document]
Comment (yuanbyu) [reply onnx/defs/rnn/defs.cc]: See above.

File: onnx/defs/rnn/defs.cc [hunk 17]
```diff
@@ -5,124 +5,311 @@
...
-
-- Equations:
-  `relu`
-  - Ht = ReLU(Wi*Xt + Ri*Ht-1 + Wbi + Rbi)
-  `tanh`
-  - Ht = tanh(Wi*Xt + Ri*Ht-1 + Wbi + Rbi)
-  `lstm`
+`H` - Hidden state
+
+Equations (LSTM with default activations):
```
[the remark this answers is not in the document]
Comment (yuanbyu) [reply onnx/defs/rnn/defs.cc]: Added the peephole term in the equation.

File: onnx/defs/rnn/defs.cc (hunk 10 above)
Comment (yuanbyu) [reply onnx/defs/rnn/defs.cc]: Good point. Done.

File: onnx/defs/rnn/defs.cc [hunk 18]
```diff
@@ -5,124 +5,311 @@
...
+	   true /*optional*/)	   
+    .Output(0, "output",
+	    "A tensor that concats all the intermediate output values of the "
+	    "hidden. It has shape `[seq_length, batch_size, hidden_size]`.")	    
+    .Output(1, "output_h",
+            "The last output value of the hidden. It has shape "
+	    "`[batch_size, hidden_size]`.");
+
+
+OPERATOR_SCHEMA(BidirectionalLSTM)
```
[the remark this answers is not in the document]
Comment (yuanbyu) [reply onnx/defs/rnn/defs.cc]: Changed to have bidirectional support for all 3 ops.

File: onnx/defs/rnn/defs.cc (hunk 11 above)
Comment (yuanbyu) [reply onnx/defs/rnn/defs.cc]: My understanding is W_i is ignored in this case. I am inclined not to create a new op for it.

**Comment by yuanbyu** [comment]
@dzhulgakov I made a round of edits based on the comments. PTAL. Thanks.

File: onnx/defs/rnn/defs.cc [hunk 19]
```diff
@@ -159,158 +164,56 @@ custom implementation such as CuDNN.
 `Wb[iofc]` - W bias vectors for input, output, forget, and cell gates
 `Rb[iofc]` - R bias vectors for input, output, forget, and cell gates
 `P[iof]`  - P peephole weight matrix for input, output, and forget gates
```
Comment (raduk) [inline onnx/defs/rnn/defs.cc]: shouldn't this be a vector ? same shape as R ?

File: onnx/defs/rnn/defs.cc [hunk 20]
```diff
@@ -6,6 +6,37 @@
...
+std::function<void(OpSchema&)> RNNDocGenerator(const char* name) {
+    return [=](OpSchema& schema) {
+        schema.Attr("hidden_size", "Number of neurons in the hidden layer", AttrType::INT);
+        schema.Attr("direction", "Specify if the RNN is forward, reverse, or bidirectional. "
+                    "Must be one of forward (default), reverse, or bidirectional.",
+                    AttrType::STRING);
+        schema.Attr("clip", "Cell clip threshold. Clipping bounds the elements of a tensor "
+                    "in the range of [-threshold, +threshold] and is applied to the input "
+                    "of activations. No clip if not specified.",
+                    AttrType::FLOAT);
```
Comment (raduk) [inline onnx/defs/rnn/defs.cc]: Coming back to this: it feels a bit weird to have the default defined by the absence. And it is not consistent with other attributes that have defaults.

Comment (raduk) [reply onnx/defs/rnn/defs.cc]: Also, from the description it may not be clear where this is applied (I've seen people mistakenly clipping after the activation for instance, or only on parts of the equation). I'd add the clip in the equations.

File: onnx/defs/rnn/defs.cc (hunk 11 above)
Comment (raduk) [reply onnx/defs/rnn/defs.cc]: I'd add this: the forget gate is computed as f = 1-i. The weights for the forget gate are ignored.

File: onnx/defs/rnn/defs.cc [hunk 21]
```diff
@@ -159,158 +164,56 @@ custom implementation such as CuDNN.
...
-    .Input(7, "seq_lens",
-           "Optional tensor specifying lengths of the sequences in a batch. "
-           "If not specified - assumed all sequences in the batch to have "
-	   "length `seq_length`. It has shape `[batch_size]`.",
+	   "to be 0. It has shape `[num_directions, batch_size, hidden_size]`.",
+	   "T", true /*optional*/)
+    .Input(7, "P",
+	   "The weight tensor for peepholes. Concatenation of `P[iof]` and "
+	   "`PB[iof]` (if bidirectional) along dimension 0. It has shape "
+	   "`[num_directions, 3*hidde_size, hidden_size]`. Optional: If not specified - "
```
Comment (raduk) [inline onnx/defs/rnn/defs.cc]: [num_directions, 3*hidden_size] ?

File: docs/Operators.md [hunk 22]
```diff
@@ -845,6 +993,82 @@
...
+  `i` - input gate
+  `o` - output gate
+  `f` - forget gate
+  `c` - cell gate
+  `t` - time step (t-1 means previous time step)
+  `W[iofc]` - W parameter weight matrix for input, output, forget, and cell gates
+  `R[iofc]` - R recurrence weight matrix for input, output, forget, and cell gates
+  `Wb[iofc]` - W bias vectors for input, output, forget, and cell gates
+  `Rb[iofc]` - R bias vectors for input, output, forget, and cell gates
+  `P[iof]`  - P peephole weight matrix for input, output, and forget gates
```
[the remark this answers is not in the document]
Comment (yuanbyu) [reply docs/Operators.md]: Good point. Fixed.

File: onnx/defs/rnn/defs.cc (hunk 21 above)
Comment (yuanbyu) [reply onnx/defs/rnn/defs.cc]: Yes. Thanks.

File: onnx/defs/rnn/defs.cc [hunk 23]
```diff
@@ -6,125 +6,214 @@
...
+`RBbi` - RR bias vectors for backward input gate
+`ReLU(X)` - max(X, 0)
+`tanh(X)` - hyperbolic tangent of X
+`H` - Hidden state
+`num_directions` - 2 if direction == bidirectional else 1
+
+Equations:
+  - Ht = Activation(Wi*Xt + Ri*Ht-1 + Wbi + Rbi)
+)DOC")
+    .Attr("activation", "The activation function for input gate. It must be "
```
Comment (raduk) [inline onnx/defs/rnn/defs.cc]: If bidirectional we need more of these

Comment (yuanbyu) [reply onnx/defs/rnn/defs.cc]: Good point. I will update the spec.

File: onnx/defs/rnn/defs.cc [hunk 24]
```diff
@@ -6,125 +6,206 @@
...
+           "Optional initial value of the cell. If not specified - assumed "
+	   "to be 0. It has shape `[num_directions, batch_size, hidden_size]`.",
+	   "T", true /*optional*/)
+    .Input(7, "P",
+	   "The weight tensor for peepholes. Concatenation of `P[iof]` and "
+	   "`PB[iof]` (if bidirectional) along dimension 0. It has shape "
+	   "`[num_directions, 3*hidde_size]`. Optional: If not specified - "
+	   "assumed to be 0.", "T",
+	   true /*optional*/)
+    .FillUsing(RNNDocGenerator("LSTM"));
```
Comment (dzhulgakov) [inline onnx/defs/rnn/defs.cc]: sorry for nitpicking, but the order of arguments is weird in this case:

X,W,R,B
initial_h
seq_lengths
initial_c
P

Maybe group them together? At least seq_len shouldn't separate out init_c and init_h. As for P it's fine as it's rarely used, so we can keep it last

File: onnx/defs/rnn/defs.cc [hunk 25]
```diff
@@ -6,125 +6,206 @@
...
+`W[iofc]` - W parameter weight matrix for input, output, forget, and cell gates
+`R[iofc]` - R recurrence weight matrix for input, output, forget, and cell gates
+`Wb[iofc]` - W bias vectors for input, output, forget, and cell gates
+`Rb[iofc]` - R bias vectors for input, output, forget, and cell gates
+`P[iof]`  - P peephole weight matrix for input, output, and forget gates
+`WB[iofc]` - W parameter weight matrix for backward input, output, forget, and cell gates
+`RB[iofc]` - R recurrence weight matrix for backward input, output, forget, and cell gates
+`WBb[iofc]` - W bias vectors for backward input, output, forget, and cell gates
+`RBb[iofc]` - R bias vectors for backward input, output, forget, and cell gates
+`PB[iof]`  - P peephole weight vector for backward input, output, and forget gates
```
Comment (dzhulgakov) [inline onnx/defs/rnn/defs.cc]: denote that it's optional

File: onnx/defs/rnn/defs.cc (hunk 24 above)
Comment (dzhulgakov) [reply onnx/defs/rnn/defs.cc]: @yuanbyu - would you mind to address this comment? 'Conditional accept' assumes that issues are addressed before merging :)

Comment (yuanbyu) [reply onnx/defs/rnn/defs.cc]: Sure. I will have a small PR for this coming your way. :-)

File: onnx/defs/rnn/defs.cc (hunk 25 above)
Comment (yuanbyu) [reply onnx/defs/rnn/defs.cc]: This is for notation. We make it clear it is optional when talking about input `P`.
